Collaboration Guide - adding challenges: release management + disagreements (#1088)
ossf/wg-best-practices-os-developers: docs/Concise-Guide-for-Collaborating-with-Open-Source-Projects.md

### End participation

> * **Wrapping up after the contribution landed:** If the contribution is of limited scope and there is no intention for a longer-term engagement with the project, clearly wrap up once the contribution has landed: regonize the support of the maintainers and close remaining open issues, such as bug reports or feature requests.

Reviewer: ```suggestion
* **Wrapping up after the contribution landed:** If the contribution is of limited scope and there is no intention for a longer-term engagement with the project, clearly wrap up once the contribution has landed. Recognize the support of the maintainers and close remaining open issues such as bug reports or feature requests.
```

> * **End long-term participation gracefully by communicating early and wrapping up:**  In case of ending a longer-term engagement, for instance due to a change in internal priorities, inform the maintainers immediately rather than simply abandoning open pull requests or discussions ("ghosting"). Make an effort to either finalize outstanding work or clearly document the state of your contribution, ensuring you do not leave the community with the burden of deciphering or maintaining unfinished code.

Reviewer: ```suggestion
* **End long-term participation gracefully by communicating early and wrapping up:**  In case of ending a longer-term engagement, for instance due to a change in internal priorities, inform the maintainers as soon as possible rather than simply abandoning open pull requests or discussions ("ghosting"). Make an effort to either finalize outstanding work or clearly document the state of your contribution, ensuring you do not leave the community with the burden of deciphering or maintaining unfinished code.
```

Sometimes company politics make it impossible to be "immediate", but you can always ask for "as soon as possible".

Larger communities, such as the CNCF, are cognizant of the need to recommend [common release practices across their projects](https://github.com/cncf/toc/issues/1849).

#### Possible Solutions

The decision to create a release remains solely at the discretion of the project maintainers. Rather than requesting an expedited release, contributors should consider the following approaches:

* Support the project maintainers in preparing the next scheduled release, for instance by running test suites and verifying release candidates.
* Contribute improvements to the project's test coverage, thereby helping to increase the maintainers' confidence in releasing more frequently.

> * Accept the responsibility of maintaining a temporary public fork, or a corresponding branch within the project's repository, until the fix is included in an official release.

Reviewer: ```suggestion
* Temporarily use a specific commit instead of a release as a dependency. Many package managers permit identifying and retrieving a specific commit instead of a formal release, specifically to deal with this situation. This avoids the problems of creating a fork. One risk is that any later security patches will not be incorporated, so monitor such situations and switch to a later commit or release when available.
* Accept the responsibility of maintaining a temporary public fork, or a corresponding branch within the project's repository, until the fix is included in an official release.
```

Author: Good suggestion!

### Disagreements about the direction of a project

> Contributions from commercial organizations are sometimes rejected by project maintainers. A maintainer may disagree with the technical approach, consider the change out of scope for the project, or determine that it introduces unwanted complexity or maintenance burden. This situation creates particular tension when the contributor feels compelled by the CRA or internal pressure to get a fix merged, while the maintainer declines for legitimate technical or architectural reasons.

Reviewer: ```suggestion
Contributions from commercial organizations are sometimes rejected by project maintainers. A maintainer may consider the change too low quality, determine that the change doesn't fully address proposed new functionality, disagree with the technical approach, consider the change out of scope for the project, or determine that it introduces unwanted complexity or maintenance burden. This situation creates particular tension when the contributor feels compelled by the CRA or internal pressure to get a fix merged, while the maintainer declines for legitimate technical or architectural reasons.
```

It is essential to recognize that maintainers have the final say over the direction of their project. Open source projects are not vendors, and maintainers are under no obligation to accept any contribution, regardless of the contributor's regulatory or business needs. Pressuring maintainers to accept changes that primarily serve commercial interests, or attempting to steer a project's roadmap to align with corporate goals, undermines the trust and collaborative spirit on which open source depends. Even when a contribution addresses a genuine security concern, the maintainer's judgment about how — or whether — to address it within their project must be respected.

#### Possible Solutions

When a contribution is rejected, contributors should consider the following approaches rather than escalating pressure on the maintainers:

> * Engage constructively with the maintainer's feedback to understand the reasons for the rejection and explore whether an alternative approach could address both the project's concerns and your requirements.

Reviewer: ```suggestion
* Engage constructively with the maintainer's feedback to understand the reasons for the rejection and explore whether an alternative approach could address both the project's concerns and your requirements. For example, if it adds new functionality but only deals with the special cases your application needs, discuss what it would take to fully address the new functionality.
```
